### PR TITLE
test(safeFloat): add extra tests

### PR DESCRIPTION
**Pull request by ttodua**

### CODE REVIEW

File: ts/src/test/base/test.safeMethods.ts [hunk 1]
```diff
@@ -385,6 +385,32 @@ function testSafeFloat () {
...
+    
+    // safeFloat - negative paths (missing key, empty string, non-numeric string, undefined container)
+    assert (exchange.safeFloat (inputDict, 'nonexistent') === undefined, 'safeFloat failed for missing key');
+    assert (exchange.safeFloat (inputDict, 'nonexistent', 5) === 5, 'safeFloat failed for missing key with default');
+    assert (exchange.safeFloat (inputDict, 'emptyString') === undefined, 'safeFloat failed for empty string');
+    assert (exchange.safeFloat (inputDict, 'str') === undefined, 'safeFloat failed for non-numeric string');
+    assert (exchange.safeFloat (inputDict, 'undefined') === undefined, 'safeFloat failed for None value');
+    assert (exchange.safeFloat (undefined, 'i') === undefined, 'safeFloat failed for undefined container');
+    assert (exchange.safeFloat (undefined, 'i', 7) === 7, 'safeFloat failed for undefined container with default');
+    assert (exchange.safeFloat (inputList, 5) === undefined, 'safeFloat failed for out-of-range list index');
```
Comment (carlotestor) [inline ts/src/test/base/test.safeMethods.ts:397]: This block covers every non-numeric input type except `bool`, which is the one that actually diverges. I ran `assert (exchange.safeFloat (inputDict, 'bool') === undefined)` through all five lanes: JS/PHP/Go return undefined/null and pass, but **Python returns `1.0`** (`safe_float` is `float(dictionary[key])`, and `float(True) == 1.0`) and **C# returns `1`** (`Convert.ToDouble` accepts `bool`). `safeString` already pins this case at line 53 — adding the `safeFloat` equivalent would catch a real 2-language divergence instead of leaving it silent. `'list'` and `'dict'` are consistent across all five and safe to add now.

File: ts/src/test/base/test.safeMethods.ts [hunk 2]
```diff
@@ -385,6 +385,32 @@ function testSafeFloat () {
     assert (exchange.safeFloatN (inputDict, [ 'a', 'b', 'strNumber' ]) === parseFloat (3));
     // @ts-expect-error
     assert (exchange.safeFloatN (inputList, [ 3, 2, 1 ]) === parseFloat (2));
+
+    
+    // safeFloat - negative paths (missing key, empty string, non-numeric string, undefined container)
+    assert (exchange.safeFloat (inputDict, 'nonexistent') === undefined, 'safeFloat failed for missing key');
+    assert (exchange.safeFloat (inputDict, 'nonexistent', 5) === 5, 'safeFloat failed for missing key with default');
```
Comment (carlotestor) [inline ts/src/test/base/test.safeMethods.ts:391]: Worth knowing before you extend this: `=== 5` works here only because every language returns the *default* unconverted (PHP keeps it `integer 5`). An assertion on a **parsed** value is different — I tried `safeFloat (inputDict, 'zeroNumeric', 99) === 0` and PHP fails it, because `safe_float` returns `float(0.0)` and PHP's `===` is type-strict (`0.0 === 0` is false). Zero-value coverage is a genuine gap (it proves `0` isn't swallowed as falsy), just assert it in a way that survives PHP's strict comparison.

**Review by carlotestor — COMMENTED** [review]
## Summary

Adds 19 negative-path assertions for the `safeFloat` / `safeFloat2` / `safeFloatN` family (missing key, empty string, non-numeric string, `undefined` value, `undefined` container, out-of-range list index, and the `default` fallback for each). Pure test additions, no source change.

I checked this out and ran it through every lane that consumes `ts/src/test/base/**`, since this directory is transpiled to Python, PHP, C# and Go — not JS-only:

| Lane | Command | Result |
|---|---|---|
| JS | `node js/src/test/base/test.safeMethods.js` | pass |
| Python | `transpile.ts --baseTests` + `test_safe_methods()` | pass |
| PHP | `php -d zend.assertions=1 php/test/tests_init.php --baseTests` | pass |
| C# | `csharpTranspiler.ts --test` + `test-base-rest-cs` | pass (build 0 warn / 0 err) |
| Go | `goTranspiler.ts --test` + `test-base-rest-go` | pass (`go build ./tests/base/` exit 0) |

All 19 new assertions transpile cleanly and hold in all five languages, so this is safe to merge as-is. The one eslint warning on the file (`11:5 unused eslint-disable directive`) is present on `origin/master` too — not from this PR.

Java doesn't pick these up: `build/javaTranspiler.ts` filters base tests on a `// AUTO_TRANSPILE_ENABLED` marker, and `test.safeMethods.ts` doesn't carry one. Pre-existing, just noting it so the coverage isn't assumed to be 7-language.

### Coverage gap worth adding while you're in here

The negative-path set covers every non-numeric input type **except boolean** — and boolean is the one that actually diverges across languages. I added `assert (exchange.safeFloat (inputDict, 'bool') === undefined)` and re-ran the full matrix:

| Lane | `safeFloat(inputDict, 'bool')` | |
|---|---|---|
| JS | `undefined` | pass |
| PHP | `NULL` | pass |
| Go | `nil` | pass |
| **Python** | **`1.0`** | **fail** |
| **C#** | **`1`** | **fail** |

Python's `safe_float` is `float(dictionary[key])` in a bare `try` (`python/ccxt/base/exchange.py:728-732`), and `float(True)` is `1.0`. C# routes through `Convert.ToDouble(result)` (`cs/ccxt/base/Exchange.SafeMethods.cs:316`), which happily converts `bool`. TS's `asFloat` gates on `isString || isNumber` (`ts/src/base/functions/type.ts:50`) so booleans fall to the default.

Note `safeString` already pins this exact case at line 53 (`safeString (inputDict, 'bool') === undefined`) — `safeFloat` is the sibling that has no such assertion. Adding it would turn a silent 2-language divergence into a caught one. Whether that lands as a test-only follow-up or a `safe_float` fix is your call; either way it's the highest-value assertion still missing here.

For reference, I also verified `list` and `dict` values are consistent (all five return undefined/null), so those two are safe to add today.

Merge gate: 🟢
Merge probability: 91%

File: ts/src/test/base/test.safeMethods.ts (hunk 1 above)
Comment (ttodua) [reply ts/src/test/base/test.safeMethods.ts:397]: needs to be fixed in python separately, will push soon

File: ts/src/test/base/test.safeMethods.ts (hunk 2 above)
Comment (ttodua) [reply ts/src/test/base/test.safeMethods.ts:391]: needs to be fixed in php separately, will push that PR too soon.

**Review by carlotestor — COMMENTED** [review]
Follow-up on 608d7da — the only delta is removing a stray trailing-whitespace line in `testSafeFloat()` between the `safeFloatN` block and the negative-path block. No assertions added, changed, or removed, so the five-lane transpile/run result from my previous review still holds unchanged.

Nothing new to flag on this commit.

Merge gate: 🟢

**Comment by carlotestor** [comment]
## Merge digest

Extends the base safe-methods test with edge-case assertions for `safeFloat`, `safeFloat2` and `safeFloatN` — missing keys, empty and non-numeric strings, undefined containers, and out-of-range list indices.

| File | Δ | What |
| --- | ---: | --- |
| `ts/src/test/base/test.safeMethods.ts` | +25/−0 | safeFloat, safeFloat2, safeFloatN edge cases |

```mermaid
flowchart LR
  testSafeFloat["testSafeFloat"]
  safeFloatN["safeFloatN"]
  safeFloat2["safeFloat2"]
  safeFloat["safeFloat"]
  testSafeFloat --> safeFloat
  testSafeFloat --> safeFloat2
  testSafeFloat --> safeFloatN
```

`ts/src/test/base/test.safeMethods.ts`

```ts
@@ -385,6 +385,31 @@
+    assert (exchange.safeFloat (inputDict, 'nonexistent') === undefined, 'safeFloat failed for missing key');
+    assert (exchange.safeFloat (inputDict, 'nonexistent', 5) === 5, 'safeFloat failed for missing key with default');
+    assert (exchange.safeFloat (inputDict, 'emptyString') === undefined, 'safeFloat failed for empty string');
+    assert (exchange.safeFloat (inputDict, 'str') === undefined, 'safeFloat failed for non-numeric string');
+    assert (exchange.safeFloat (inputDict, 'undefined') === undefined, 'safeFloat failed for None value');
+    assert (exchange.safeFloat (undefined, 'i') === undefined, 'safeFloat failed for undefined container');
+    assert (exchange.safeFloat (undefined, 'i', 7) === 7, 'safeFloat failed for undefined container with default');
+    assert (exchange.safeFloat (inputList, 5) === undefined, 'safeFloat failed for out-of-range list index');
+    assert (exchange.safeFloat2 (inputDict, 'nonexistent', 'nonexistent2') === undefined, 'safeFloat2 failed for missing keys');
+    assert (exchange.safeFloat2 (inputDict, 'nonexistent', 'str') === undefined, 'safeFloat2 failed for missing then non-numeric');
+    assert (exchange.safeFloat2 (inputDict, 'nonexistent', 'emptyString') === undefined, 'safeFloat2 failed for missing then empty string');
+    assert (exchange.safeFloat2 (inputDict, 'nonexistent', 'nonexistent2', 9) === 9, 'safeFloat2 failed for missing keys with default');
+    assert (exchange.safeFloat2 (undefined, 'i', 'f') === undefined, 'safeFloat2 failed for undefined container');
```
